ostree/ostree_src.go

```diff
@@ -0,0 +1,168 @@
+	imgspecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+type ostreeImageSource struct {
+	ref    ostreeReference
+	tmpDir string
+}
+
+// newImageSource returns an ImageSource for reading from an existing directory.
+func newImageSource(ctx *types.SystemContext, tmpDir string, ref ostreeReference) (types.ImageSource, error) {
+	return &ostreeImageSource{ref: ref, tmpDir : tmpDir}, nil
+}
+
+// Reference returns the reference used to set up this source.
+func (s *ostreeImageSource) Reference() types.ImageReference {
+	return s.ref
+}
+
+// Close removes resources associated with an initialized ImageSource, if any.
+func (s *ostreeImageSource) Close() error {
+	return nil
+}
+
+func (s *ostreeImageSource) getLayerSize(blob string) (int64, error) {
+	b := fmt.Sprintf("ociimage/%s", blob)
+	// Use golang bindings once they support to read metadata
+	out, err := exec.Command("ostree", "show", "--repo", s.ref.repo, "--print-metadata-key=docker.size", b).CombinedOutput()
+	if err != nil {
+		fmt.Printf("FAIL1\n")
+		return 0, err
+	}
+
+	size, err := strconv.ParseInt(string(out[1:len(out)-2]), 10, 64)
+	if err != nil {
+		return 0, err
+	}
+	return size, nil
+}
+
+func (s *ostreeImageSource) getTarSplitData(blob string) ([]byte, error) {
+	b := fmt.Sprintf("ociimage/%s", blob)
+	// Use golang bindings once they support to read metadata
+	out, err := exec.Command("ostree", "show", "--repo", s.ref.repo, "--print-metadata-key=tarsplit.output", b).CombinedOutput()
+	if err != nil {
+		if strings.Index(string(out), "No such metadata key") >= 0 {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return base64.StdEncoding.DecodeString(string(out[1:len(out)-2]))
+
+}
+
+// GetManifest returns the image's manifest along with its MIME type (which may be empty when it can't be determined but the manifest is available).
+// It may use a remote (= slow) service.
+func (s *ostreeImageSource) GetManifest() ([]byte, string, error) {
+	b := fmt.Sprintf("ociimage/%s", s.ref.branchName)
+	// Use golang bindings once they support to read metadata
+	out, err := exec.Command("ostree", "show", "--repo", s.ref.repo, "--print-metadata-key=docker.manifest", b).Output()
+	if err != nil {
+		return nil, "", err
+	}
+	m := []byte(strings.Replace(string(out[1:len(out)-2]), "\\n", "\n", -1))
+	return m, manifest.GuessMIMEType(m), nil
+}
+
+func (s *ostreeImageSource) GetTargetManifest(digest digest.Digest) ([]byte, string, error) {
+	return nil, imgspecv1.MediaTypeImageManifest, nil
+}
+
+func (s *ostreeImageSource) checkout(dest string, checksum string) error {
+	opts := otbuiltin.NewCheckoutOptions()
+	opts.UserMode = true
+	commit := fmt.Sprintf("ociimage/%s", checksum)
+	return otbuiltin.Checkout(s.ref.repo, dest, commit, opts)
+}
+
+// GetBlob returns a stream for the specified blob, and the blob's size.
+func (s *ostreeImageSource) GetBlob(info types.BlobInfo) (io.ReadCloser, int64, error) {
+	blob := strings.Replace(string(info.Digest), "sha256:", "", 1)
```

Comment:
Should this be `info.Digest.Hex()`? Or is any non-`sha256` digest supposed to be preserved including the prefix?

Reply:
yes, fixed now.